WordPress/gutenberg
# Global styles: background UI controls

## Dev Note 📓 

https://github.com/WordPress/gutenberg/pull/59354#issuecomment-2138583222

## What?

Part of:

- https://github.com/WordPress/gutenberg/issues/54336

Follow up to:

- https://github.com/WordPress/gutenberg/pull/59354

Related:

- https://github.com/WordPress/gutenberg/pull/60100

This PR add background image UI controls to top-level global styles. 

### TODO

- Maybe E2E tests...

## Why?

Background styles can now be added to theme.json since https://github.com/WordPress/gutenberg/pull/59354.

The styles are currently only " top-level", which means they're added to `body` selector. 

This PR allows users to edit these styles in the site editor.

## How?

Similar to dimensions and typography panels, this PR refactors the background block supports component so that it can be shared between the site and post editors.

It moves all relevant tests and styles to the `/global-styles` directory.

## Testing Instructions

This is a big one, so get comfortable.

Firstly, check out this branch.

At this stage it's probably good to check if there are any regressions to background block supports. So, add a group block to a post, and make sure you can:

1. Add/remove/upload/reset a new background image (See https://github.com/WordPress/gutenberg/pull/53934 for test instructions)
2. Edit/reset the background position values (https://github.com/WordPress/gutenberg/pull/58592)
3. Edit/reset background size and repeat values (https://github.com/WordPress/gutenberg/pull/57005)

If that's all good, head over to the site editor.

I've been using empty theme to test. Here is some test theme.json, with a relative path to the updated theme.json schema:

```json
{
	"$schema": "../../schemas/json/theme.json",
	"version": 2,
	"settings": {
		"appearanceTools": true,
		"background": {
			"backgroundImage": true,
			"backgroundSize": true
		}
	},
	"styles": {
		"color": {
			"background": "tomato"
		},
		"background": {
			"backgroundImage": {
				"url": "https://i0.wp.com/wordpress.org/files/2023/12/sotw-dotorg-drawer.png?w=1807&ssl=1",
				"source": "file"
			},
			"backgroundPosition": "30% 30%",
			"backgroundRepeat": "repeat",
			"backgroundSize": "cover"
		}
	}
}
```

Ensure that:

1. Any inherited values from theme.json are present.
2. You can interact with the background controls in a similar way to above, e.g., at the block level, and
3. Saving the styles affects the body styles of your site. 
4. Also check that the background size defaults are: "cover" for Group blocks, and `undefined` for site backgrounds. See table images below.


| Group block background controls | Top-level global styles background controls |
|--------|--------|
| <img width="288" alt="Screenshot 2024-03-21 at 3 25 15 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/586ea0d1-c69d-4194-a469-4c93709f4e24"> | <img width="288" alt="Screenshot 2024-03-21 at 3 24 56 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/99d2fe79-f3fe-4bfc-9a07-df6ab415d6be"> | 


### Testing Instructions for Keyboard

In the global styles panel, ensure you can navigate to the "Background" item in the list to activate the background styles panel. There should be no regressions to keyboard interaction with background controls.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/6458278/8b4274e5-f0a9-4ab0-b34d-dc8d2e200255
